### PR TITLE
Added `metadata` and `visibility` to the organisational unit model

### PR DESCRIPTION
**Pull request by simong**

### CODE REVIEW

File: node_modules/gh-core/lib/db.js [hunk 1]
```diff
@@ -354,10 +354,12 @@ var _setUpModel = function(sequelize) {
...
      *
-     * @property  {Number}      id                  The unique identifier of the organizational unit
-     * @property  {String}      displayName         The display name of the organizational unit
-     * @property  {Group}       group               The group who can edit this organizational unit and the units below it
-     * @property  {String}      type                A label that identifies what this organization unit represents within the tree of organizational units
+     * @property  {Number}      id                  The unique identifier of the organisational unit
+     * @property  {String}      displayName         The display name of the organisational unit
+     * @property  {Group}       group               The group who can edit this organisational unit and the units below it
+     * @property  {String}      type                A label that identifies what this organization unit represents within the tree of organisational units
+     * @property  {String}      visibility          A label that identifies the visibility status of the organisational unit. One of `draft` or `published`
```
Comment (nicolaasmatthijs) [inline node_modules/gh-core/lib/db.js]: Should this just be a `draft` boolean?

File: node_modules/gh-core/lib/db.js [hunk 2]
```diff
@@ -354,10 +354,12 @@ var _setUpModel = function(sequelize) {
...
+     * @property  {Object}      metadata            An object with extra metadate for the organisational unit
```
Comment (nicolaasmatthijs) [inline node_modules/gh-core/lib/db.js]: typo

Comment (nicolaasmatthijs) [inline node_modules/gh-core/lib/db.js]: Should also indicate that this is optional

File: node_modules/gh-orgunit/lib/api.js [hunk 3]
```diff
@@ -41,15 +41,18 @@ var OrgUnitDAO = require('./internal/dao');
...
  * @param  {Number}         [parentId]          The id of the parent organisational unit. By not providing a parent id, you will end up with a top level organisational unit
  * @param  {Function}       callback            Standard callback function
  * @param  {Object}         callback.err        An error object, if any
  * @param  {OrgUnit}        callback.orgUnit    The created organisational unit
  */
-var createOrgUnit = module.exports.createOrgUnit = function(ctx, appId, displayName, type, description, groupId, parentId, callback) {
+var createOrgUnit = module.exports.createOrgUnit = function(ctx, appId, displayName, type, description, metadata, visibility, groupId, parentId, callback) {
     // Ensure that the app id is a valid number
     appId = GrasshopperUtil.getNumberParam(appId);
+    metadata = metadata || {};
```
Comment (nicolaasmatthijs) [inline node_modules/gh-orgunit/lib/api.js]: We should probably also default the `draft` flag to `false`

File: node_modules/gh-orgunit/lib/api.js [hunk 4]
```diff
@@ -69,6 +72,20 @@ var createOrgUnit = module.exports.createOrgUnit = function(ctx, appId, displayN
         validator.check(parentId, {'code': 400, 'msg': 'A valid parent organisational unit id must be provided'}).isInt();
         parentId = GrasshopperUtil.getNumberParam(parentId);
     }
+    if (metadata) {
+        if (_.isString(metadata)) {
+            try {
```
Comment (nicolaasmatthijs) [inline node_modules/gh-orgunit/lib/api.js]: Maybe we should update the JSDocs and specify that both `string` and `object` are accepted

File: node_modules/gh-orgunit/lib/rest.js [hunk 5]
```diff
@@ -271,9 +275,11 @@ GrassHopper.globalAdminRouter.on('post', '/api/orgunit', createOrgUnit);
  * @PathParam   {number}            id                  The id of the organisational unit to update
  * @FormParam   {string}            [description]       Updated organisational unit description
  * @FormParam   {string}            [displayName]       Updated organisational unit name
- * @FormParam   {group}             [group]             Updated id of the group that can manage the organisational unit
+ * @FormParam   {number}            [group]             Updated id of the group that can manage the organisational unit
+ * @FormParam   {string}            [metadata]          The metadata of the organisational unit
```
Comment (nicolaasmatthijs) [inline node_modules/gh-orgunit/lib/rest.js]: Should specify that this should be a JSON string

File: node_modules/gh-orgunit/lib/restmodel.js [hunk 6]
```diff
@@ -19,16 +19,25 @@
...
  * @Property  {string}          type                The organisational unit type (e.g. `tripos`, `part`)
  * @Property  {string}          updatedAt           The timestamp (ISO 8601) at which the organisational unit was last updated
+ * @Property  {string}          visibility          The visibility of the organisational unit
+ */
+
+/**
+ * @RESTModel OrgUnitUpdate
+ *
+ * @Required  [{key}]
+ * @Property  {string}         {key}            Some metadata
```
Comment (nicolaasmatthijs) [inline node_modules/gh-orgunit/lib/restmodel.js]: I'm assuming this can be removed?

File: node_modules/gh-orgunit/tests/test-orgunit.js [hunk 7]
```diff
@@ -134,6 +147,36 @@ describe('Organisational units', function() {
...
+
+                // Verify a simple metadata object with multiple keys
+                opts.metadata = {
+                    'url': 'http:/www.google.com',
+                    'name': 'Simon',
+                };
+                OrgUnitTestUtil.assertCreateOrgUnit(global.tests.admins.cam2013.client, 'Test orgunit', 'course', opts, function(orgUnit) {
+
+                    // Verify nested objects
+                    opts.metadata = {
```
Comment (nicolaasmatthijs) [inline node_modules/gh-orgunit/tests/test-orgunit.js]: Maybe also include an array?

File: node_modules/gh-rest/lib/rest/orgunit.js [hunk 8]
```diff
@@ -87,6 +87,8 @@ module.exports = function(client) {
             'displayName': displayName,
             'group': opts.group,
             'parent': opts.parent,
+            'metadata': opts.metadata,
```
Comment (nicolaasmatthijs) [inline node_modules/gh-rest/lib/rest/orgunit.js]: Should update JSDocs here

**Comment by nicolaasmatthijs** [comment]
Finished review. Back to @simong for follow-up

**Comment by simong** [comment]
Done, back to @nicolaasmatthijs

File: node_modules/gh-core/lib/db.js [hunk 9]
```diff
@@ -369,6 +371,14 @@ var _setUpModel = function(sequelize) {
             'allowNull': false
         },
         'description': Sequelize.STRING,
+        'metadata': {
+            'type': Sequelize.JSON,
+            'allowNull': true
+        },
+        'published': {
+            'type': Sequelize.BOOLEAN,
+            'allowNull': true
```
Comment (nicolaasmatthijs) [inline node_modules/gh-core/lib/db.js]: Should we set the default to `true` ?

File: node_modules/gh-orgunit/lib/api.js [hunk 10]
```diff
@@ -41,15 +41,18 @@ var OrgUnitDAO = require('./internal/dao');
  * @param  {String}         displayName         The name of the organisational unit
  * @param  {String}         type                The organisational unit type (e.g. `tripos`, `part`)
  * @param  {String}         [description]       The description of the organisational unit
+ * @param  {Object|String}  [metadata]          The extra metadata for the organisational unit
```
Comment (nicolaasmatthijs) [inline node_modules/gh-orgunit/lib/api.js]: Should specify that when using a string, it should be a stringified JSON object

**Comment by nicolaasmatthijs** [comment]
Finished review. Back to @simong for follow-up

**Comment by simong** [comment]
Done. Back to @nicolaasmatthijs

**Comment by nicolaasmatthijs** [comment]
Looks good. Merged.

**Pull request closed without merging**